apic_ml2/neutron/db/l3out_vlan_allocation.py

```diff
@@ -0,0 +1,241 @@
+
+    When an allocation is released, if the vlan_id is inside the pool
+    described by network_vlan_ranges, then allocated is set to
+    False. If it is outside the pool, the record is deleted.
+    """
+
+    __tablename__ = 'apic_ml2_l3out_vlan_allocation'
+    __table_args__ = (
+        sa.Index('apic_ml2_l3out_vlan_allocation_l3out_network_allocated',
+                 'l3out_network', 'allocated'),
+        model_base.BASEV2.__table_args__,)
+
+    l3out_network = sa.Column(sa.String(64), nullable=False,
+                              primary_key=True)
+    vrf = sa.Column(sa.String(64), nullable=False,
+                    primary_key=False)
+    vlan_id = sa.Column(sa.Integer, nullable=False, primary_key=True,
+                        autoincrement=False)
+    allocated = sa.Column(sa.Boolean, nullable=False)
+
+
+class NoVlanAvailable(exc.ResourceExhausted):
+    message = _("Unable to allocate the vlan. "
+                "No vlan is available for %(l3out_network)s external network")
+
+
+# inherit from SegmentTypeDriver to reuse the code to reserve/release
+# vlan IDs from the pool
+class L3outVlanAlloc(helpers.SegmentTypeDriver):
+
+    def __init__(self):
+        super(L3outVlanAlloc, self).__init__(L3OutVlanAllocation)
+        self.session = db_api.get_session()
+
+    def _parse_vlan_ranges(self, ext_net_dict):
+        self.l3out_vlan_ranges = {}
+        try:
+            for l3out_network in ext_net_dict.keys():
+                ext_info = ext_net_dict.get(l3out_network)
+                vlan_range_str = ext_info.get('shadow_l3out_vlan_range')
+                if vlan_range_str:
+                    vlan_min, vlan_max = vlan_range_str.strip().split(':')
+                    vlan_range = (int(vlan_min), int(vlan_max))
+                    plugin_utils.verify_vlan_range(vlan_range)
+                    self.l3out_vlan_ranges[l3out_network] = vlan_range
+        except Exception:
+            LOG.exception(_LE("Failed to parse L3out_vlan_ranges."))
+            raise
+        LOG.info(_LI("L3out VLAN ranges: %s"), self.l3out_vlan_ranges)
+
+    def sync_vlan_allocations(self, ext_net_dict):
+        self._parse_vlan_ranges(ext_net_dict)
+        with self.session.begin(subtransactions=True):
+            # get existing allocations for all L3 out networks
+            allocations = dict()
+            allocs = (self.session.query(L3OutVlanAllocation).
+                      with_lockmode('update'))
+            for alloc in allocs:
+                if alloc.l3out_network not in allocations:
+                    allocations[alloc.l3out_network] = set()
+                allocations[alloc.l3out_network].add(alloc)
+
+            # process vlan ranges for each configured l3out network
+            for (l3out_network,
+                 vlan_range) in self.l3out_vlan_ranges.items():
+                # determine current configured allocatable vlans for
+                # this l3out network
+                vlan_ids = set()
```

Comment:
turned out this line is needed. Otherwise it will complain "referenced before assignment" error.....

raceback (most recent call last):
  File "apic_ml2/neutron/tests/unit/db/test_l3out_vlan_allocation.py", line 47, in setUp
    self.l3out_vlan_alloc.sync_vlan_allocations(ext_net_dict)
  File "apic_ml2/neutron/db/l3out_vlan_allocation.py", line 106, in sync_vlan_allocations
    vlan_ids |= set(moves.xrange(vlan_min, vlan_max + 1))
UnboundLocalError: local variable 'vlan_ids' referenced before assignment